src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1755,206 +1906,299 @@ public static decimal Sum<TSource>(this IEnumerable<TSource> source, Func<TSourc
 
         public static int Min(this IEnumerable<int> source)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            int value = 0;
-            bool hasValue = false;
-            foreach (int x in source)
+            Error.CheckSourceNotNull(source);
+            int value;
+            using (IEnumerator<int> e = source.GetEnumerator())
             {
-                if (hasValue)
+                if (!e.MoveNext()) throw Error.NoElements();
+                value = e.Current;
+                while (e.MoveNext())
                 {
+                    int x = e.Current;
                     if (x < value) value = x;
                 }
-                else
-                {
-                    value = x;
-                    hasValue = true;
-                }
             }
-            if (hasValue) return value;
-            throw Error.NoElements();
+            return value;
```

Comment:
Very hard to answer in general, because some of the branches are more likely to be mis-predicted than others.
Some of the changes to Min allow for a short-circuit, specifically those on floating-point values, which is both the biggest possible gain for some sequences in this whole PR (this whole thing was inspired by a bug I had a while back with how Single doesn't short-circuit) and the thing that could cause issues if people are expecting side-effects (if you are coding side-effects into your Linq use you are going to have a bad time, but that doesn't mean people won't have done so). I haven't added short-circuits where there isn't a test in the original, but it's still at least 5 cases of short-circuiting that are new. Perhaps whether or not to short-circuit needs to be discussed separately.